### PR TITLE
Add method for adding lazy values to AttributeContainer using providers 

### PR DESCRIPTION
**Pull request by tresat**
This work was extracted from the PR 18834.

### Contributor Checklist
- [ ] [Review Contribution Guidelines](https://github.com/gradle/gradle/blob/master/CONTRIBUTING.md)
- [ ] Make sure that all commits are [signed off](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---signoff) to indicate that you agree to the terms of [Developer Certificate of Origin](https://developercertificate.org/).
- [ ] Make sure all contributed code can be distributed under the terms of the [Apache License 2.0](https://github.com/gradle/gradle/blob/master/LICENSE), e.g. the code was written by yourself or the original code is licensed under [a license compatible to Apache License 2.0](https://apache.org/legal/resolved.html).
- [ ] Check ["Allow edit from maintainers" option](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/) in pull request so that additional changes can be pushed by Gradle team
- [ ] Provide integration tests (under `<subproject>/src/integTest`) to verify changes from a user perspective
- [ ] Provide unit tests (under `<subproject>/src/test`) to verify logic
- [ ] Update User Guide, DSL Reference, and Javadoc for public-facing changes
- [ ] Ensure that tests pass sanity check: `./gradlew sanityCheck`
- [ ] Ensure that tests pass locally: `./gradlew <changed-subproject>:quickTest`

### Gradle Core Team Checklist
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation
- [ ] Recognize contributor in release notes

### CODE REVIEW

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@big-guy, @DPUkyle - Here is just the AttributeContainer work lifted from https://github.com/gradle/gradle/pull/18834, using the generic type arg vs. the 2-arg call, and with an upgrade note added.  Hopefully this looks good to merge soon.

**Comment by tresat** [comment]
@bot-gradle test this

File: build-logic-commons/code-quality/src/main/kotlin/gradlebuild.code-quality.gradle.kts [hunk 1]
```diff
@@ -41,7 +41,7 @@ val rules by configurations.creating {
     isCanBeConsumed = false
 
     attributes {
-        attribute(LibraryElements.LIBRARY_ELEMENTS_ATTRIBUTE, objects.named(LibraryElements.RESOURCES))
+        attribute(LibraryElements.LIBRARY_ELEMENTS_ATTRIBUTE, objects.named<LibraryElements>(LibraryElements.RESOURCES))
```
Comment (DPUkyle) [inline build-logic-commons/code-quality/src/main/kotlin/gradlebuild.code-quality.gradle.kts:44]: Nevermind, I see you explained it in `upgrading_version_7.adoc`. Are we really sure Kotlin can't handle both argument types? Perhaps not, if the use-case is Object vs. Provider.

--

I wonder why we started parameterizing all the `objects.named` calls. I don't have a problem with it, I'm just curious.

Is it something to do with the erasure of `AttributeContainer#attribute(Attribute<T>, T)` conflicting with `public <T> AttributeContainer attribute(Attribute<T> key, Provider<? extends T> provider)` on `DefaultImmutableAtributes`?

File: subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java [hunk 2]
```diff
@@ -46,23 +51,34 @@ public String toString() {
...
+    public <T> AttributeContainer attribute(Attribute<T> key, Provider<? extends T> provider) {
+        assertAttributeTypeIsValid(provider, key);
+        checkInsertionAllowed(key);
+        lazyAttributes.put(key, provider);
         return this;
     }
 
     private <T> void checkInsertionAllowed(Attribute<T> key) {
-        for (Attribute<?> attribute : state.keySet()) {
+        // Don't just use keySet() method instead, since we should be allowed to override attributes alread in the parent
```
Comment (DPUkyle) [inline subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java:79]: 👍

File: subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java [hunk 3]
```diff
@@ -138,16 +197,31 @@ public boolean equals(Object o) {
...
         return result;
     }
+
+    private ImmutableAttributes evaluateLazyValues() {
+        final AttributeContainerInternal evaluatedAttributes = immutableAttributesFactory.mutable();
+        for (Map.Entry<Attribute<?>, Provider<?>> entry : lazyAttributes.entrySet()) {
+            @SuppressWarnings("unchecked") Attribute<Object> attribute = (Attribute<Object>) entry.getKey();
+            Object value = entry.getValue().get();
+            evaluatedAttributes.attribute(attribute, value);
+        }
```
Comment (DPUkyle) [inline subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java:224]: See https://github.com/gradle/gradle/pull/18834#discussion_r746973844

**Review by DPUkyle — CHANGES_REQUESTED** [review]
Overall looks good, with two points for follow-up:

 1. I am not a Kotlin expert. Are we 100% sure the new method overload breaks the kotlin extension function?
 2. Consider using `Map#forEach` instead of iterating over entryset

File: subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java [hunk 4]
```diff
@@ -20,22 +20,27 @@
...
 import java.util.Map;
+import java.util.Objects;
 import java.util.Set;
 
 class DefaultMutableAttributeContainer implements AttributeContainerInternal {
-    private final ImmutableAttributesFactory cache;
+    private final ImmutableAttributesFactory immutableAttributesFactory;
     private final AttributeContainerInternal parent;
     private ImmutableAttributes state = ImmutableAttributes.EMPTY;
+    private final Map<Attribute<?>, Provider<?>> lazyAttributes = Maps.newHashMap();
```
Comment (DPUkyle) [inline subprojects/core/src/main/java/org/gradle/api/internal/attributes/DefaultMutableAttributeContainer.java:36]: Please use `new HashMap<>();` for the RHS. I suspect this should also remove the need to use a cast in the `forEach` use below.

From the javadoc of `Maps#newHashMap()`:
```
Note for Java 7 and later: this method is now unnecessary and should be treated as deprecated. Instead, use the HashMap constructor directly, taking advantage of the new "diamond" syntax .
```

Even without the "soft" deprecation above, I think we should not add any more usages of guava.

**Review by DPUkyle — CHANGES_REQUESTED** [review]

**Comment by tresat** [comment]
@bot-gradle test this

**Comment by tresat** [comment]
@big-guy, @DPUkyle - I think I've addressed everything for the AttributeContainer work here and this is ready to merge.

**Review by big-guy — APPROVED** [review]

**Comment by big-guy** [comment]
@bot-gradle test and merge

**Comment by big-guy** [comment]
@bot-gradle cancel

**Comment by big-guy** [comment]
@bot-gradle test and merge

**Comment by big-guy** [comment]
@bot-gradle test and merge

**Pull request merged**